invoices/invoiceregistry_test.go

```diff
@@ -947,3 +947,50 @@ func TestInvoiceExpiryWithRegistry(t *testing.T) {
 		}
 	}
 }
+
+// TestProcessKeysend asserts the behavior of processKeySend as it relates to
+// generating and inserting invoices into the invoice database.
+func TestProcessKeysend(t *testing.T) {
+	defer timeout()()
+
+	ctx := newTestContext(t)
+	defer ctx.cleanup()
+
+	const height = 500
+
+	preimage := lntypes.Preimage{1}
+	keysendCtx := invoiceUpdateCtx{
+		hash:                 preimage.Hash(),
+		circuitKey:           channeldb.CircuitKey{HtlcID: 1},
+		amtPaid:              1,
+		currentHeight:        height,
+		expiry:               uint32(height + testFinalCltvRejectDelta + 1),
+		finalCltvRejectDelta: testFinalCltvRejectDelta,
+		customRecords: map[uint64][]byte{
+			record.KeySendType: preimage[:],
+		},
+		mpp: nil,
+	}
+
+	// Inserting a new keysend payment should succeed.
+	err := ctx.registry.processKeySend(keysendCtx)
+	assert.Nil(t, err)
+
+	// Replaying the same context should also succeed to ensure idempotency.
+	// This should be handled later in the pipeline by comparing the circuit
+	// keys.
+	err = ctx.registry.processKeySend(keysendCtx)
```

Comment:
I think that testing on the external interface of invoice registry is more robust.

Reply:
my thought here was that this is a better place to assert the behavior of `processKeySend` wrt to how it filters errors

Reply:
Why is it better? We also already have a parameterized `TestKeysend` that tests on the external interface. Maybe that can be extended easily.

Reply:
afaict you can't test the idempotency of `processKeySend` because it settles the keysend immediately after

Reply:
moved everything to a white box test, but we are discussing holding off on this change and fixing as part of #4338